lnst/Tests/XDPBench.py

```diff
@@ -0,0 +1,139 @@
+import re
+import time
+import signal
+import logging
+from subprocess import Popen, PIPE
+from lnst.Devices.Device import Device
+
+from lnst.Tests.BaseTestModule import BaseTestModule, TestModuleError
+from lnst.Common.Parameters import (
+    ChoiceParam,
+    StrParam,
+    IntParam,
+    DeviceParam,
+)
+
+
+class XDPBenchOutputParser:
+    def __init__(self, process: Popen, output: list) -> None:
+        self._process = process
+        self._out = output
+
+    def parse_output(self):
+        stdout, stderr = self._process.communicate()
+
+        logging.debug("Stderr of xdp-bench:")
+        logging.debug(str(stderr))
+
+        self._parse_raw_results(stdout)
+
+        if not self._out:
+            raise TestModuleError(f"Could not parse xdp-bench output: {stdout}")
+
+    def _parse_raw_results(self, raw_results: bytes):
+        decoded = raw_results.decode()
+        intervals = decoded.splitlines()
+
+        for line in intervals:
+            self._parse_line(line)
+
+    def _parse_line(self, line: str):
+        match = re.search(r"Summary\s+([\d,]+)\srx/s\s+([\d,]+)\serr/s?", line)
+
+        if not match:  # skip summary line at the end + corrupted lines
+            logging.error(f"Could not parse xdp-bench output: {line}")
+            return
+
+        rx = match.group(1).replace(",", "")
+        err = match.group(2).replace(",", "")
+        # ^^ remove thousands separators
+
+        self._out.append({"rx": int(rx), "err": int(err)})
+
+XDP_BENCH_COMMANDS = (
+    "pass",
+    "drop",
+    "tx",
+    "redirect",
+    "redirect-cpu",
+    "redirect-map",
+    "redirect-multi",
+)
+
+class XDPBench(BaseTestModule):
+    """
+    xdp-bench tool abstraction. [1]
+
+    This tool does NOT check params validity.
+
+    xdp-bench is expected to be included in PATH env variable.
+
+    [1] https://github.com/xdp-project/xdp-tools/
+    """
+
+    command = ChoiceParam(
+        type=StrParam,
+        choices=XDP_BENCH_COMMANDS,
+        mandatory=True,
+    )
+    interface = DeviceParam(mandatory=True)
+    interface2 = DeviceParam()  # used for redirect modes
+
+    interval = IntParam(default=1)
+
+    redirect_device = DeviceParam()
```

Comment:
Isn't this the same as `interface2` as it mentions redirect?

Reply:
Well, it might be, but having it as a 2 separate params makes the implementation of bash command composing much simpler. `interface2` param is used for `xdp-bench` modes that doesn't require argument name e.g. `./xdp-bench redirect-map ingress_inf egress_inf`. However mode `redirect-cpu` requires argument `--redirect-device inf`.

I didn't want to dig into command composing based on further context, see https://github.com/LNST-project/lnst/blob/0a0afe2fe0ed24144b8101b00089234f44d95dca/lnst/Tests/XDPBench.py#L120  so i decided to include both arguments there.

Reply:
Ok, if the tool takes this as separate arguments, then that's ok.